File: scripts/build_pack_manifest.py
#!/usr/bin/env python3
"""build_pack_manifest.py — (re)generate a pack's pack-manifest.json.

Python port of backend/scripts/build-pack-manifest.js for the sync
pipeline (which runs where Node may not be installed). The two builders
MUST stay byte-identical: tests/test_build_pack_manifest.py proves this
transitively by asserting this builder reproduces the COMMITTED
manifests, which backend/tests/contract/pack/pack-schemas.test.js pins
to the Node builder (the manifest-freshness contract test).

Machine-maintains the INVENTORY half of the manifest (files[] with
sha1+size, and contentHash over the sorted "path:sha1" lines) while
PRESERVING the hand-authored half (packId, version, engine, hardware,
createdAt) from an existing manifest. A pack with no manifest yet gets a
skeleton with TODO hardware that fails schema validation until authored —
intentional: the hardware section is a design statement, never derivable.

Usage: python3 build_pack_manifest.py <packDir>
"""

import hashlib
import json
import re
import sys
from pathlib import Path

EXCLUDE = {
    "pack-manifest.json",
    "game.schema.json",
    "pack-manifest.schema.json",
    "tokens.schema.json",
    "scoring-config.json",  # legacy shared file — retired by the migration, never pack inventory
    "CLAUDE.md",
    "README.md",
}

SKIP_DIRS = {"node_modules", "shared"}


def role_for(rel_path):
    if rel_path == "game.json":
        return "game"
    if rel_path == "tokens.json":
        return "tokens"
    if rel_path == "strings.json":
        return "strings"
    if rel_path == "theme.json":
        return "theme"
    if rel_path == "cues.json":
        return "cues"
    if rel_path.startswith("templates/"):
        return "template"
    if re.match(r"^assets/images/", rel_path):
        return "asset-image"
    if re.match(r"^assets/audio/", rel_path):
        return "asset-audio"
    if re.search(r"\.mp4$", rel_path):
        return "asset-video"
    return "other"


def _walk(directory, base, out):
    for entry in sorted(directory.iterdir()):
        if entry.name.startswith("."):
            continue
        rel = entry.relative_to(base).as_posix()
        if entry.is_dir():
            if entry.name in SKIP_DIRS:
                continue
            _walk(entry, base, out)
        elif entry.is_file():
            if rel in EXCLUDE:
                continue
            if rel.endswith(".html"):
                continue
            out.append(rel)
    return out


def build_files(pack_dir):
    pack_dir = Path(pack_dir)
    files = []
    for rel in sorted(_walk(pack_dir, pack_dir, [])):
        buf = (pack_dir / rel).read_bytes()
        files.append({
            "path": rel,
            "role": role_for(rel),
            "sha1": hashlib.sha1(buf).hexdigest(),
            "size": len(buf),
        })
    return files


def content_hash(files):
    lines = "\n".join(sorted(f"{f['path']}:{f['sha1']}" for f in files))
    return f"sha256:{hashlib.sha256(lines.encode('utf-8')).hexdigest()}"


def build(pack_dir):
    """Build the manifest dict for pack_dir (does not write).

    Key ORDER matters: it must match the Node builder exactly for byte
    parity of the serialized output.
    """
    pack_dir = Path(pack_dir)
    manifest_path = pack_dir / "pack-manifest.json"
    existing = (
        json.loads(manifest_path.read_text(encoding="utf-8"))
        if manifest_path.exists()
        else {}
    )

    files = build_files(pack_dir)
    manifest = {
        "kind": "pack-manifest",
        "schemaVersion": 1,
        "packId": existing.get("packId") or pack_dir.name,
        "version": existing.get("version") or "0.1.0",
        "contentHash": content_hash(files),
    }
    if existing.get("createdAt"):
        manifest["createdAt"] = existing["createdAt"]
    manifest["engine"] = existing.get("engine") or {"minVersion": "3.0.0"}
    manifest["files"] = files
    manifest["hardware"] = existing.get("hardware") or {
        "deviceClasses": [
            {"class": "staffed", "min": 1, "rationale": "TODO: author the hardware section"}
        ],
    }
    return manifest, manifest_path


def serialize(manifest):
    """Byte-identical to the Node builder's JSON.stringify(m, null, 2) + '\\n'."""
    return json.dumps(manifest, indent=2, ensure_ascii=False) + "\n"


def write_manifest(pack_dir):
    manifest, manifest_path = build(pack_dir)
    manifest_path.write_text(serialize(manifest), encoding="utf-8")
    return manifest, manifest_path


if __name__ == "__main__":
    if len(sys.argv) != 2 or not Path(sys.argv[1]).is_dir():
        print("Usage: python3 build_pack_manifest.py <packDir>", file=sys.stderr)
        sys.exit(1)
    m, p = write_manifest(Path(sys.argv[1]).resolve())
    print(f"{p}: {len(m['files'])} files, {m['contentHash'][:23]}…")

File: backend/scripts/build-pack-manifest.js
#!/usr/bin/env node
/**
 * build-pack-manifest.js — (re)generate a pack's pack-manifest.json
 *
 * Usage: node scripts/build-pack-manifest.js <packDir>
 *
 * Machine-maintains the INVENTORY half of the manifest (files[] with
 * sha1+size, and contentHash over the sorted "path:sha1" lines) while
 * PRESERVING the hand-authored half (packId, version, engine, hardware,
 * createdAt) from an existing manifest. A pack with no manifest yet gets a
 * skeleton with TODO hardware that fails schema validation until authored —
 * intentional: the hardware section is a design statement, never derivable.
 *
 * File roles are inferred from path conventions; unknown files get 'other'.
 * The manifest itself and non-pack repo files (schemas, docs, dotfiles) are
 * excluded from the inventory.
 *
 * Deterministic: same tree in = byte-identical files[]/contentHash out
 * (the freshness contract test relies on this).
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const EXCLUDE = new Set([
  'pack-manifest.json',
  'game.schema.json',
  'pack-manifest.schema.json',
  'tokens.schema.json',
  'CLAUDE.md',
  'README.md',
]);

function roleFor(relPath) {
  if (relPath === 'game.json') return 'game';
  if (relPath === 'tokens.json') return 'tokens';
  if (relPath === 'strings.json') return 'strings';
  if (relPath === 'theme.json') return 'theme';
  if (relPath === 'cues.json') return 'cues';
  if (relPath.startsWith('templates/')) return 'template';
  if (/^assets\/images\//.test(relPath)) return 'asset-image';
  if (/^assets\/audio\//.test(relPath)) return 'asset-audio';
  if (/\.mp4$/.test(relPath)) return 'asset-video';
  return 'other';
}

function walk(dir, base = dir, out = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    if (entry.name.startsWith('.')) continue;
    const full = path.join(dir, entry.name);
    const rel = path.relative(base, full).split(path.sep).join('/');
    if (entry.isDirectory()) {
      if (entry.name === 'node_modules' || entry.name === 'shared') continue;
      walk(full, base, out);
    } else if (entry.isFile()) {
      if (EXCLUDE.has(rel)) continue;
      // Non-pack top-level extras (HTML utilities etc.) are inventoried as
      // 'other' only when clearly content; skip known repo tooling
      if (/\.(html)$/.test(rel)) continue;
      out.push(rel);
    }
  }
  return out;
}

function buildFiles(packDir) {
  return walk(packDir)
    .sort()
    .map((rel) => {
      const buf = fs.readFileSync(path.join(packDir, rel));
      return {
        path: rel,
        role: roleFor(rel),
        sha1: crypto.createHash('sha1').update(buf).digest('hex'),
        size: buf.length,
      };
    });
}

function contentHash(files) {
  const lines = files.map((f) => `${f.path}:${f.sha1}`).sort().join('\n');
  return `sha256:${crypto.createHash('sha256').update(lines).digest('hex')}`;
}

function build(packDir) {
  const manifestPath = path.join(packDir, 'pack-manifest.json');
  const existing = fs.existsSync(manifestPath)
    ? JSON.parse(fs.readFileSync(manifestPath, 'utf8'))
    : {};

  const files = buildFiles(packDir);
  const manifest = {
    kind: 'pack-manifest',
    schemaVersion: 1,
    packId: existing.packId || path.basename(packDir),
    version: existing.version || '0.1.0',
    contentHash: contentHash(files),
    ...(existing.createdAt ? { createdAt: existing.createdAt } : {}),
    engine: existing.engine || { minVersion: '3.0.0' },
    files,
    hardware: existing.hardware || {
      deviceClasses: [{ class: 'staffed', min: 1, rationale: 'TODO: author the hardware section' }],
    },
  };
  return { manifest, manifestPath };
}

if (require.main === module) {
  const packDir = process.argv[2];
  if (!packDir || !fs.existsSync(packDir)) {
    console.error('Usage: node scripts/build-pack-manifest.js <packDir>');
    process.exit(1);
  }
  const { manifest, manifestPath } = build(path.resolve(packDir));
  fs.writeFileSync(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`);
  console.log(`${manifestPath}: ${manifest.files.length} files, ${manifest.contentHash.slice(0, 23)}…`);
}

module.exports = { build, buildFiles, contentHash };
